### PR TITLE
[4.0] Pluginless lazyloading for the core

### PR DESCRIPTION
**Pull request by zero-24**
Pull Request for Issue #30218 

### Summary of Changes

This is an replacement of https://github.com/joomla/joomla-cms/pull/30218 by dropping the core lazyloading plugin ~~as well as implementing higth & width adttributes to the core fields~~.

~~Based on @dgrammatiko 's JS code I hope i did not broke anything in that area?~~

Stoke through parts implemented with PR #30784 .

## clean install test

- Install the installation package from here:  [https://ci.joomla.org/artifacts/joomla/joomla-cms/4.0-dev/30748/downloads/35816/](https://ci.joomla.org/artifacts/joomla/joomla-cms/4.0-dev/30748/downloads/35816/)
- make sure that there is no lazyload plugin any more

## upgrade test
1. On a clean 4.0-dev or 4.0 Beta 4 or latest 4.0 nightly installation, update to the update package built by Drone for this PR. You can find an update package and a custom update URL here: [https://ci.joomla.org/artifacts/joomla/joomla-cms/4.0-dev/30748/downloads/35816/](https://ci.joomla.org/artifacts/joomla/joomla-cms/4.0-dev/30748/downloads/35816/).
2. Verify that the update suceeds without SQL errors.
3. Check in the list of plugins in backend that the plugin can't be found anymore.

### Actual result BEFORE applying this Pull Request

We set the loading attribute without the width and heigth attribute.

### Expected result AFTER applying this Pull Request

There is no lazyloading plugin any more ~~but still we set the loading attribute for new images alongside the width and heigth attributes~~.

### Documentation Changes Required

None

### Question 

Should there be a option to opt-out of automatic lazyloading new images?
I'm not sure whether we might want to document the SQL scripts to remove the lazyloading plugin from the database? @richard67 @wilsonge ?

### CODE REVIEW

**Comment by dgrammatiko** [comment]
@zero-24 the images rendered through the media field still have no width and height but easily fixable:
change https://github.com/joomla/joomla-cms/blob/621171ff2ca3739f4923efb7c7204074cb93113d/build/media_source/system/js/fields/joomla-field-media.w-c.es6.js#L43
to 
```js
editor.value = `${Joomla.selectedFile.url}?width=${Joomla.selectedFile.width}&height=${Joomla.selectedFile.height}`;
```

Explainer: The value that is stored in the db for each form field media **is** a **url**. Urls can have params. We're passing the required data as params, and pick them up in the layout [php: parse_url($url)] and append them to the respected image attributes. FWIW this technic was proposed to solve also the editing of an image (by having a ?version=xxx param) but was never implemented in the new media manager due to the lack of db integration (I think that was planned for the next iteration).

**Comment by zero-24** [comment]
Patched thanks @dgrammatiko

**Comment by Quy** [comment]
Close #30615?

File: installation/sql/mysql/base.sql [hunk 1]
```diff
@@ -363,7 +363,6 @@ INSERT INTO `#__extensions` (`package_id`, `name`, `type`, `element`, `folder`,
 (0, 'plg_workflow_publishing', 'plugin', 'publishing', 'workflow', 0, 1, 1, 0, 1, '', '{}', '', 0, 0),
 (0, 'plg_workflow_featuring', 'plugin', 'featuring', 'workflow', 0, 1, 1, 0, 1, '', '{}', '', 0, 0),
 (0, 'plg_workflow_notification', 'plugin', 'notification', 'workflow', 0, 1, 1, 0, 1, '', '{}', '', 0, 0),
```
Comment (Quy) [inline installation/sql/mysql/base.sql]:
```suggestion
(0, 'plg_workflow_notification', 'plugin', 'notification', 'workflow', 0, 1, 1, 0, 1, '', '{}', '', 0, 0);
```

File: installation/sql/postgresql/base.sql [hunk 2]
```diff
@@ -369,7 +369,6 @@ INSERT INTO "#__extensions" ("package_id", "name", "type", "element", "folder",
 (0, 'plg_workflow_publishing', 'plugin', 'publishing', 'workflow', 0, 1, 1, 0, 1, '', '{}', '', 0, 0),
 (0, 'plg_workflow_featuring', 'plugin', 'featuring', 'workflow', 0, 1, 1, 0, 1, '', '{}', '', 0, 0),
 (0, 'plg_workflow_notification', 'plugin', 'notification', 'workflow', 0, 1, 1, 0, 1, '', '{}', '', 0, 0),
```
Comment (Quy) [inline installation/sql/postgresql/base.sql]:
```suggestion
(0, 'plg_workflow_notification', 'plugin', 'notification', 'workflow', 0, 1, 1, 0, 1, '', '{}', '', 0, 0);
```

**Comment by zero-24** [comment]
> Close #30615?

Done

File: installation/sql/mysql/base.sql (hunk 1 above)
Comment (richard67) [reply installation/sql/mysql/base.sql]: @zero-24 Please apply this suggestion, otherwise installation on MySQL fails, which currently is the case for the MySQL system tests in Drone.

Comment (zero-24) [reply installation/sql/mysql/base.sql]: I thourgth i already did that :D Now it is applyed

Comment (richard67) [reply installation/sql/mysql/base.sql]: The "already did" was maybe the other file for postgresql ;-)

**Comment by dgrammatiko** [comment]
@zero-24 it seems that I broke something in the media manager at some point. Please add
```js 
                                width: this.item.width ? this.item.width : undefined,
                                height: this.item.height? this.item.height : undefined,
```
after https://github.com/joomla/joomla-cms/blob/319edca39ec47ebff646cc13e42a2637b1b62b93/administrator/components/com_media/resources/scripts/components/browser/items/item.js#L95

Also here's a quick implementation of fulltext image layout:
```php
defined('_JEXEC') or die;

use Joomla\Utilities\ArrayHelper;

$params  = $displayData->params;
$images  = json_decode($displayData->images);
$image   = parse_url($images->image_fulltext);
$src     = $image['path'];
$attr    = [];

parse_str($image['query'], $imageParams);

if (count($imageParams))
{
	if ($imageParams['width'] !== 'undefined')
	{
		$attr['width'] = $imageParams['width'] . 'px';
	}

	if ($imageParams['height'] !== 'undefined')
	{
		$attr['height'] = $imageParams['height'] . 'px';
	}
}
?>
<?php if (!empty($images->image_fulltext)) : ?>
	<?php $imgfloat = empty($images->float_fulltext) ? $params->get('float_fulltext') : $images->float_fulltext; ?>
	<figure class="float-<?php echo htmlspecialchars($imgfloat, ENT_COMPAT, 'UTF-8'); ?> item-image">
		<img loading="lazy" src="<?php echo htmlspecialchars($src, ENT_COMPAT, 'UTF-8'); ?>"
			 alt="<?php echo htmlspecialchars($images->image_fulltext_alt, ENT_COMPAT, 'UTF-8'); ?>"
			 itemprop="image"
				<?php echo ArrayHelper::toString($attr); ?>/>
		<?php if (isset($images->image_fulltext_caption) && $images->image_fulltext_caption !== '') : ?>
			<figcaption class="caption"><?php echo htmlspecialchars($images->image_fulltext_caption, ENT_COMPAT, 'UTF-8'); ?></figcaption>
		<?php endif; ?>
	</figure>
<?php endif;
```

and a screenshot of the expected result
Screenshot 2020-09-23 at 20 26 12

**Comment by zero-24** [comment]
Thanks @dgrammatiko i have just implemended the suggested changes for the full and intro image layout as well as the mention JS changes in the media manager. Thanks!

**Comment by SharkyKZ** [comment]
🤦‍♂️

**Comment by zero-24** [comment]
What is wrong @SharkyKZ ?

**Comment by SharkyKZ** [comment]
Everything. Revert previous commits and don't screw around with image URLs please.

**Comment by zero-24** [comment]
> Everything. Revert previous commits and don't screw around with image URLs please.

You mean the layout changes right? what is wrong with them? We can also branch this out to a dedicated PR when we need to cosider more things on that.

**Comment by dgrammatiko** [comment]
> Everything. Revert previous commits and don't screw around with image URLs please.

I'm curious as well can you elaborate please?

**Comment by SharkyKZ** [comment]
Not that anyone cares, but Media Manager is supposed to work with external image hosts.

**Comment by dgrammatiko** [comment]
> Not that anyone cares, but Media Manager is supposed to work with external image hosts.

It does, there's an API for that

FYI: https://github.com/joomla/joomla-cms/blob/319edca39ec47ebff646cc13e42a2637b1b62b93/administrator/components/com_media/src/Adapter/AdapterInterface.php#L21-L43
All adapters should return width and height.
Also the PHP part in the layout, as I wrote, is a quick implementation mostly for testing purposes. Pretty sure won't take much effort to cover everything

File: build/media_source/system/js/fields/joomla-field-media.w-c.es6.js [hunk 3]
```diff
@@ -68,7 +68,7 @@
     const apiBaseUrl = `${Joomla.getOptions('system.paths').baseFull}index.php?option=com_media&format=json`;
 
     Joomla.request({
-      url: `${apiBaseUrl}&task=api.files&url=true&path=${data.path}&${Joomla.getOptions('csrf.token')}=1&format=json`,
+      url: `${apiBaseUrl}&task=api.files&url=true&path=${data.path}&${Joomla.getOptions('csrf.token')}=1`,
```
Comment (brianteeman) [inline build/media_source/system/js/fields/joomla-field-media.w-c.es6.js]: just checking it was deliberate to remove format=json

Comment (zero-24) [reply build/media_source/system/js/fields/joomla-field-media.w-c.es6.js]: yes when you take a look at line 68 there is the base path defined that already includes format=json ;)

File: layouts/joomla/content/full_image.php [hunk 4]
```diff
@@ -9,17 +9,43 @@
...
+use Joomla\Utilities\ArrayHelper;
+
+$params   = $displayData->params;
+$images   = json_decode($displayData->images);
+$imgfloat = empty($images->float_fulltext) ? $params->get('float_fulltext') : $images->float_fulltext;
+$image    = parse_url($images->image_fulltext);
+$src      = $image['path'];
+$attr     = [];
+
+parse_str($image['query'], $imageParams);
```
Comment (Quy) [inline layouts/joomla/content/full_image.php]: > Undefined index: query

File: layouts/joomla/content/full_image.php [hunk 5]
```diff
@@ -9,17 +9,43 @@
...
 defined('_JEXEC') or die;
 
-$params  = $displayData->params;
-$images  = json_decode($displayData->images);
+use Joomla\Utilities\ArrayHelper;
+
+$params   = $displayData->params;
+$images   = json_decode($displayData->images);
+$imgfloat = empty($images->float_fulltext) ? $params->get('float_fulltext') : $images->float_fulltext;
+$image    = parse_url($images->image_fulltext);
```
Comment (Quy) [inline layouts/joomla/content/full_image.php]: > Notice: Trying to get property 'image_fulltext' of non-object

Comment (dgrammatiko) [reply layouts/joomla/content/full_image.php]: My bad, the code needs to be wrapped in an if statement as it happens later on on the html

Comment (dgrammatiko) [reply layouts/joomla/content/full_image.php]: here's the fulltext updated:
```php
defined('_JEXEC') or die;

use Joomla\Utilities\ArrayHelper;

$params = $displayData->params;
$images = json_decode($displayData->images);

if (!empty($images->image_fulltext)) {
	$image = parse_url($images->image_fulltext);
	$attr  = [];

	parse_str($image['query'], $imageParams);

	if (count($imageParams))
	{
		if ($imageParams['width'] !== 'undefined')
		{
			$attr['width'] = $imageParams['width'] . 'px';
		}

		if ($imageParams['height'] !== 'undefined')
		{
			$attr['height'] = $imageParams['height'] . 'px';
		}
	}
}
?>
<?php if (!empty($images->image_fulltext)) : ?>
	<?php $imgfloat = empty($images->float_fulltext) ? $params->get('float_fulltext') : $images->float_fulltext; ?>
	<figure class="float-<?php echo htmlspecialchars($imgfloat, ENT_COMPAT, 'UTF-8'); ?> item-image">
		<img loading="lazy" src="<?php echo htmlspecialchars($images->image_fulltext, ENT_COMPAT, 'UTF-8'); ?>"
			 alt="<?php echo htmlspecialchars($images->image_fulltext_alt, ENT_COMPAT, 'UTF-8'); ?>"
			 itemprop="image"
				<?php echo ArrayHelper::toString($attr); ?>/>
		<?php if (isset($images->image_fulltext_caption) && $images->image_fulltext_caption !== '') : ?>
			<figcaption class="caption"><?php echo htmlspecialchars($images->image_fulltext_caption, ENT_COMPAT, 'UTF-8'); ?></figcaption>
		<?php endif; ?>
	</figure>
<?php endif;
```

and here's the intro:
```php
defined('_JEXEC') or die;

use Joomla\CMS\Router\Route;
use Joomla\Component\Content\Site\Helper\RouteHelper;
use Joomla\Utilities\ArrayHelper;

$params   = $displayData->params;
$images   = json_decode($displayData->images);
$imgfloat = empty($images->float_intro) ? $params->get('float_intro') : $images->float_intro;

if (!empty($images->image_intro)) {
	$image = parse_url($images->image_intro);
	$attr  = [];

	parse_str($image['query'], $imageParams);

	if (count($imageParams))
	{
		if ($imageParams['width'] !== 'undefined')
		{
			$attr['width'] = $imageParams['width'] . 'px';
		}

		if ($imageParams['height'] !== 'undefined')
		{
			$attr['height'] = $imageParams['height'] . 'px';
		}
	}
}

?>
<?php if (!empty($images->image_intro)) : ?>
	<figure class="float-<?php echo htmlspecialchars($imgfloat, ENT_COMPAT, 'UTF-8'); ?> item-image">
		<?php if ($params->get('link_titles') && $params->get('access-view')) : ?>
			<a href="<?php echo Route::_(RouteHelper::getArticleRoute($displayData->slug, $displayData->catid, $displayData->language)); ?>">
				<img
						loading="lazy"
						src="<?php echo htmlspecialchars($images->image_intro, ENT_COMPAT, 'UTF-8'); ?>"
						alt="<?php echo htmlspecialchars($images->image_intro_alt, ENT_COMPAT, 'UTF-8'); ?>"
						itemprop="thumbnailUrl"
						<?php echo ArrayHelper::toString($attr); ?>
				/>
			</a>
		<?php else : ?>
			<img
					loading="lazy"
					src="<?php echo htmlspecialchars($images->image_intro, ENT_COMPAT, 'UTF-8'); ?>"
					alt="<?php echo htmlspecialchars($images->image_intro_alt, ENT_COMPAT, 'UTF-8'); ?>"
					itemprop="thumbnailUrl"
					<?php echo ArrayHelper::toString($attr); ?>
			/>
		<?php endif; ?>
		<?php if (isset($images->image_intro_caption) && !empty($images->image_intro_caption)) : ?>
			<figcaption class="caption">
				<?php echo htmlspecialchars($images->image_intro_caption, ENT_COMPAT, 'UTF-8'); ?>
			</figcaption>
		<?php endif; ?>
	</figure>
<?php endif; ?>
```

@SharkyKZ I guess these changes bring us back to the previous state (no destruction/reconstruction of the url)
@zero-24 sorry for the buggy code, could you update the 2 layouts?

Comment (zero-24) [reply layouts/joomla/content/full_image.php]: No problems this should be even easier as we don't have to repeate the if's right?: https://github.com/joomla/joomla-cms/pull/30748/commits/8bc50053a37ce44577237d959b6db47a3c1cee76

Comment (SharkyKZ) [reply layouts/joomla/content/full_image.php]: You're still appending random query parameters to the URL which is wrong. It may or may not cause issues depending on image host.

Comment (SharkyKZ) [reply layouts/joomla/content/full_image.php]: @zero-24 Image check has to be after `json_decode()`.

Comment (dgrammatiko) [reply layouts/joomla/content/full_image.php]: >  It may or may not cause issues depending on image host.

That's a bold claim. If there were to be any problems with any query params would be on the client side but all browsers know how to handle urls. It's the backbone of the internet...

Comment (zero-24) [reply layouts/joomla/content/full_image.php]: Right done: https://github.com/joomla/joomla-cms/pull/30748/commits/50ce77dafd067c6a7716a4ca3ccb8f6458ef32e5

File: layouts/joomla/content/full_image.php [hunk 6]
```diff
@@ -9,17 +9,47 @@
...
+$attr     = [];
+
+parse_str($image['query'], $imageParams);
+
+if (count($imageParams))
+{
+	if ($imageParams['width'] !== 'undefined')
+	{
+		$attr['width'] = $imageParams['width'] . 'px';
```
Comment (Quy) [inline layouts/joomla/content/full_image.php]: Not necessary to add `px`.

Comment (zero-24) [reply layouts/joomla/content/full_image.php]: Fixed by: https://github.com/joomla/joomla-cms/pull/30748/commits/720ae43bab3c6cce95f98bd3469ad0d7f10c25bf

File: layouts/joomla/content/full_image.php (hunk 5 above)
Comment (SharkyKZ) [reply layouts/joomla/content/full_image.php]: @dgrammatiko You are totally missing the point. Query strings could be used by the host one way or another. We can't possibly know.

**Comment by dgrammatiko** [comment]
Since @SharkyKZ raised some questions about the proposal to pass the additional data (width and height) as query params in the image url, I think I should try to debunk those and explain a bit more why this is possibly the only viable way.

> It may or may not cause issues depending on image host.

Joomla is appending query parameters in CSS and JS files for almost one decade without anyone (at least as far as I know) complaining that their host cannot handle these properly. I don't see how a Joomla host will be broken on any image URL that has some extra query params.

> But J4's media manager opens the door for external storages. 

Sure, and the API already cover us here as it's mandatory to return the image width and height in the Adapter.

> So the API is fine but what about the actual host?

Well, Dropbox, S3, Cloudinary, etc. (almost every popular service I know of) will handle the extra URL params. Most of them are already using URL params one way or another. But if that's still a concern there's a way out, keep reading

> How about name classing?

There are 3 ways out of this:
- destructure and re construct of the URL removing the extra URL params (I've tried that but was very rough)
- use unique naming, eg `joomla_img_width` and `joomla_img_height`
- both destructure and re construct of the URL and unique naming 👍

> How about B/C break here?

There is no possible B/C break by adding a couple of params in the URL. That said I want to mention that the new media manager supports more storages other than the localhost. So, if some developers have specific code that treats these URLs as part of the localhost path (eg concatenate with `JPATH_ROOT' . $url`) that will not work, and this is **not** something introduced here, it's far deeper (the Adapter part, but these changes make it way more obvious as they also affect the localhost).

> Can't we just do some simple PHP check on the file meta?

No, we can't  do that for two main reasons:
- It will not work for any other storage other than localhost
- The performance difference is significant here (having to read each image per request to get the dimensions hardly passes and perf consideration)
(Not to mention that is good to have one point of truth, eg the database)

@zero-24 one more change in the js, please replace https://github.com/joomla/joomla-cms/blob/720ae43bab3c6cce95f98bd3469ad0d7f10c25bf/build/media_source/system/js/fields/joomla-field-media.w-c.es6.js#L37-L46 with:

```js
      const appendParam = (url, key, value) => {
        const newKey = encodeURIComponent(key);
        const newValue = encodeURIComponent(value);
        const r = new RegExp(`(&|\\?)${key}=[^\&]*`);
        let s = url;
        const param = `${newKey}=${newValue}`;

        s = s.replace(r, `$1${param}`);

        if (!RegExp.$1 && s.includes('?')) {
          return `${s}&${param}`;
        }

        if (!RegExp.$1 && !s.includes('?')) {
          return `${s}?${param}`;
        }

        return s;
      };

      if (Joomla.selectedFile.url) {
        if (!isElement(editor) && (typeof editor !== 'object')) {
          Joomla.editors.instances[editor].replaceSelection(`<img loading="lazy" src="${Joomla.selectedFile.url}" width="${Joomla.selectedFile.width}" height="${Joomla.selectedFile.height}" alt="" />`);
        } else if (!isElement(editor) && (typeof editor === 'object' && editor.id)) {
          window.parent.Joomla.editors.instances[editor.id].replaceSelection(`<img loading="lazy" src="${Joomla.selectedFile.url}" width="${Joomla.selectedFile.width}" height="${Joomla.selectedFile.height}" alt="" />`);
        } else {
          const val = appendParam(Joomla.selectedFile.url, 'joomla_image_width', Joomla.selectedFile.width);
          editor.value = appendParam(val, 'joomla_image_height', Joomla.selectedFile.height);
          fieldClass.updatePreview();
        }
      }
```
Basically it will respect any preexisting URL params, previous code was not treating correctly existing URL params.

PS 1. Please mind the change in the names of the URL params

PS 2 . About the restructuring and reconstructing the URL, these can be done as an HTMLHelper and the code is already widely available, eg: https://stackoverflow.com/questions/4354904/php-parse-url-reverse-parsed-url (destructuring is supported by PHP so all it might need is return an object with all the parts or false if it fails)

**Comment by richard67** [comment]
@zero-24 As we support updates from Beta to later versions (Beta or RC or final) in J4, should the plugin be uninstalled when updating? Maybe you can discuss that directly with George? I'm not sure (yet) how we should handle that.

**Comment by zero-24** [comment]
> @zero-24 As we support updates from Beta to later versions (Beta or RC or final) in J4, should the plugin be uninstalled when updating? Maybe you can discuss that directly with George? I'm not sure (yet) how we should handle that.

Yes that is the same thing i had in mind too. I dont want to implement beta to beta SQL or thing to the script PHP. But lucky us nothing should break in the end so a note that on upgrade there might should be some cleanup script executed would be ok? @wilsonge ?

> @zero-24 one more change in the js, please replace

 @dgrammatiko just to be sure in that JS code we would change to `joomla_image_width` and `joomla_image_heigth` so we should do that on all places as well as edit all the other code to make sure they check for the new values right?

**Comment by richard67** [comment]
@zero-24 Maybe we will be "lucky" and something else breaks update to the next beta, then we don't have to do anything anyway 😉 But you are right, if we do something it will be script.php, using model, othewise we won't get rid of stuff like assets.

**Comment by dgrammatiko** [comment]
> should do that on all places as well as edit all the other code to make sure they check for the new values right

Yup, it's the naming on the layouts, intro and fulltext. Also if you think that snake_case is not the preferred case please change  it to whatever is more appropriate (camelCase, pascal, kebab). Finally you might want to rename the names to something else (naming is hard)

**Comment by zero-24** [comment]
Thanks yes i have just added that to my list hopefully i get this cleandup by next week, thanks for all that feedback! :)

**Comment by dgrammatiko** [comment]
@zero-24 can you rollback to only uninstall the plugin here? 
I already prepared a PR that has all the needed JS, layout, etc https://github.com/joomla/joomla-cms/pull/30784

**Comment by zero-24** [comment]
Fine for me will do that hopefully tomorrow Thabks for your PR!

**Comment by zero-24** [comment]
This PR has now been reset to only remove the plugin and add the is_array check for the HTML Helper. Thanks @dgrammatiko 👍

**Comment by dgrammatiko** [comment]
I have tested this item :white_check_mark: successfully on 8de20811e1280ec79f0a39c146b3cddb54e5c6a5<br /><br />@zero-24 Thank you for this one!<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30748">issues.joomla.org/tracker/joomla-cms/30748</a>.</sub>

**Comment by Quy** [comment]
I have tested this item :white_check_mark: successfully on 8de20811e1280ec79f0a39c146b3cddb54e5c6a5<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30748">issues.joomla.org/tracker/joomla-cms/30748</a>.</sub>

**Comment by Quy** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30748">issues.joomla.org/tracker/joomla-cms/30748</a>.</sub>

File: libraries/src/HTML/HTMLHelper.php [hunk 7]
```diff
@@ -688,7 +688,7 @@ public static function image($file, $alt, $attribs = null, $relative = false, $r
 		}
 
 		// Default to lazy you can disable lazyloading by passing $attribs['loading'] = 'eager';
-		if (!isset($attribs['loading']))
+		if (is_array($attribs) && !isset($attribs['loading']))
```
Comment (SharkyKZ) [inline libraries/src/HTML/HTMLHelper.php]: This should really be removed. But if it stays, it should at least work the same regardless of whether `$attribs` is an array or not.

Comment (zero-24) [reply libraries/src/HTML/HTMLHelper.php]:
```suggestion
		if (!isset($attribs['loading']))
```

Comment (zero-24) [reply libraries/src/HTML/HTMLHelper.php]: I have just reverted the change from here so this here can be mergedn and i will do a new PR with support for the string and array syntax

**Comment by richard67** [comment]
@dgrammatiko @Quy We (Tobias and me) had to add an additional test, see section "upgrade test" in the testing instructions. Could you test that, too? The other test and related code has not been changed, so no need to repeat your previous, valid test for the "clean install test" section. Thanks in advance.

**Review by richard67 — APPROVED** [review]

**Comment by dgrammatiko** [comment]
I have tested this item :white_check_mark: successfully on e7350599f6a835413439ba67a75c389fef441186<br /><br />Update of a clean J4 was fine, plugin was removed successfully <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30748">issues.joomla.org/tracker/joomla-cms/30748</a>.</sub>

**Comment by richard67** [comment]
Update test on MySQLi and PostgreSQL is ok. Now am doing new installation tests for both database types.

**Comment by richard67** [comment]
I have tested this item :white_check_mark: successfully on e7350599f6a835413439ba67a75c389fef441186<br /><br />
- New installation MySQLi: :white_check_mark:
- New installation PostgreSQL (PDO): :white_check_mark:
- Update MySQLi: :white_check_mark:
- Update PostgreSQL (PDO): :white_check_mark:<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30748">issues.joomla.org/tracker/joomla-cms/30748</a>.</sub>

File: libraries/src/HTML/HTMLHelper.php (hunk 7 above)
Comment (zero-24) [reply libraries/src/HTML/HTMLHelper.php]: Done here: https://github.com/joomla/joomla-cms/pull/30790

**Comment by HLeithner** [comment]
Thanks for doing it the proper way.

**Comment by zero-24** [comment]
Thanks
